### ISSUE TITLE
Filter does not work with Identity attribute as specified

### ISSUE THREAD

**Issue by sarollins** [MicrosoftDocs/office-docs-powershell#9043]
In anticipation of the -LdapFilter being deprecated, I'm attempting to re-write some of our queries to get a large subset of users, by using the -Filter parameter instead

When I attempt to use Identity in the Filter (Get-CsOnlineUser -Filter {Identity -eq d3c4b4e1-b747-4372-ac51-c9d7f2d13620}), I get the error

Get-CsOnlineUser : Cannot bind parameter 'Filter' to the target. Exception setting "Filter": "Query not supported for
property: "Identity" query: "Identity -eq d3c4b4e1-b747-4372-ac51-c9d7f2d13620" position: "1""

Ultimately, I need to replicate the functionality we are able to use in the -LdapFilter parameter (because we have clients with over 100,000 users, we have to get their data in chunks of users, so we cycle through the 2 digit hex values 00 - FF to get all the users)

Get-CsOnlineUser -LdapFilter "(objectGUID=\00*)"

So, I would expect the syntax to be something like:

Get-CsOnlineUser -Filter {Identity -like '00*'}

Piping the output of Get-CsOnlineUser to Where does work, but the performance is ultimately too slow for our clients

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 7008ec10-a056-d8bb-bb4e-b4690a84adfd
* Version Independent ID: 2dcb78d9-7f36-208e-0989-184905d73011
* Content: [Get-CsOnlineUser (SkypeForBusiness)](https://docs.microsoft.com/en-us/powershell/module/skype/get-csonlineuser?view=skype-ps)
* Content Source: [skype/skype-ps/skype/Get-CsOnlineUser.md](https://github.com/MicrosoftDocs/office-docs-powershell/blob/main/skype/skype-ps/skype/Get-CsOnlineUser.md)
* Service: **skypeforbusiness-powershell**
* GitHub Login: @tomkau
* Microsoft Alias: **tomkau**

**Comment by dariomws** [comment]
Hi @sarollins, thank you for your feedback.

It works on my side:
<img width="449" alt="image" src="https://user-images.githubusercontent.com/33589238/159575135-8c4b5ff4-78d5-47f6-b6b5-09fc3df991f3.png">
<img width="880" alt="image" src="https://user-images.githubusercontent.com/33589238/159575185-16b36337-19cb-435a-b8dc-6dd7ef78c827.png">

Please try the latest version (4.0.0) of the Teams PowerShell module if you are not using it.

**Comment by sarollins** [comment]
Yes, 4.0.0 does work, if a single identity is specified

***@***.***

But my ultimate goal of getting large chunks of users using the like command does not work

***@***.***

**Comment by dariomws** [comment]
@isrumnon, could you kindly help here?

It seems that the Filter parameter is not working as expected for Identity:
<img width="1266" alt="image" src="https://user-images.githubusercontent.com/33589238/159579293-f9000f12-7329-4992-adfc-2e1c8cfc9c94.png">

**Comment by isrumnon** [comment]
Hi @dariomws I'm looking into this and will update you shortly

**Comment by yogkumgit** [comment]
@isrumnon Please update and help us in resolving this issue. Thanks

**Comment by sarollins** [comment]
I still see issues even running version 4.1.0 of the MicrosoftTeams module

**Comment by isrumnon** [comment]
Can you please check again now and see if the command works? 
@sarollins can you summarize the issues you are facing?

**Comment by sarollins** [comment]
Some great strides have been made, but as you see, Enabled is a deprecated attribute and should not have a value, whereas the opposite is true for IsSipEnabled

PS D:\docs\ps\sfbo> Get-CsOnlineUser -Filter {Identity -eq 'd3c4b4e1-b747-4372-ac51-c9d7f2d13620'} | Select AccountEnabled,Enabled,IsSipEnabled

AccountEnabled Enabled IsSipEnabled
-------------- ------- ------------
          True                False

And this still does not work (which is our ultimate goal)

PS D:\docs\ps\sfbo> Get-CsOnlineUser -Filter {Identity -like 'd3*'}
Get-CsOnlineUser : AADGraph Request failed with BadRequest. AAD Code  Message .
Parameter name: https://graph.windows.net/a5c32d43-5d6f-41a1-a3da-1390153869b6/users?api-version=1.61-internal&$select=
assignedPlans,cloudMSRtcIsSipEnabled,preferredLanguage,shadowProxyAddresses,displayName,jobTitle,givenName,surname,acco
untEnabled,streetAddress,cloudSipLine,cloudSipProxyAddress,state,userPrincipalName,proxyAddresses,objectId,usageLocatio
n,dirSyncEnabled,UserType,country,department,provisionedPlans,cloudMSRtcOwnerUrn,tenantId,cloudMSRtcServiceAttributes,c
loudMSRtcPolicyAssignments,mailNickname,companyName,postalCode,city&$top=100&$filter=(startswith(ObjectId, %27d3%27))
At line:1 char:1
+ Get-CsOnlineUser -Filter {Identity -like 'd3*'}
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidOperation: ({ PageSize = , ...et = Extended }:<>f__AnonymousType81`11) [Get-CsOnl
   ineUser], Exception
    + FullyQualifiedErrorId : BadArgument,Microsoft.Teams.ConfigApi.Cmdlets.GetCsOnlineUser

And I am running 4.1.0 of the MicrosoftTeams module

**Comment by yogkumgit** [comment]
@isrumnon

**Comment by sarollins** [comment]
Running 4.2.0 of the MicrosoftTeams module, I still get the same error

**Comment by isrumnon** [comment]
HI @sarollins 

- Enabled is now renamed to IsSipEnabled hence you might be seeing this error

**Comment by sarollins** [comment]
I'm not sure what the IsSipEnabled attribute has to do with this not working

Get-CsOnlineUser -Filter {Identity -like 'd3*'}

Get-CsOnlineUser : AADGraph Request failed with BadRequest. AAD Code  Message .
Parameter name: https://graph.windows.net/a5c32d43-5d6f-41a1-a3da-1390153869b6/users?api-version=1.61-internal&$select=
assignedPlans,cloudMSRtcIsSipEnabled,preferredLanguage,shadowProxyAddresses,displayName,jobTitle,givenName,surname,acco
untEnabled,streetAddress,cloudSipLine,cloudSipProxyAddress,state,userPrincipalName,proxyAddresses,objectId,usageLocatio
n,dirSyncEnabled,UserType,country,department,provisionedPlans,cloudMSRtcOwnerUrn,tenantId,cloudMSRtcServiceAttributes,c
loudMSRtcPolicyAssignments,mailNickname,companyName,postalCode,city,cloudMSRtcPool&$top=100&$filter=(startswith(ObjectI
d, %27d3%27))
At line:1 char:1
+ Get-CsOnlineUser -Filter {Identity -like 'd3*'}
    + CategoryInfo          : InvalidOperation: ({ PageSize = , ...et = Extended }:<>f__AnonymousType81`11) [Get-CsOnl
   ineUser], Exception
    + FullyQualifiedErrorId : BadArgument,Microsoft.Teams.ConfigApi.Cmdlets.GetCsOnlineUser

**Comment by isrumnon** [comment]
Hi @sarollins filtering based on Identity in conjunction with the -Like operator is currently not supported. 
You can try possible alternatives like filtering based on alias or upn?

**Comment by sarollins** [comment]
Microsoft is putting us (and our customers) in a bind

As per the documentation for Get-CsOnlineUser, the -LdapFilter parameter “has been deprecated from the Teams Powershell Modules version 3.0 or later as it is no longer relevant to Microsoft Teams.”

First off, I’m running 4.3.0 and this parameter is still valid (thank goodness)

We deal with clients who have a lot of users (some over 200K).  Due to PowerShell memory issues, it is not possible for us to get all users and necessary attributes in a single call.  So, we have to get the users in chunks.  Since we also have international customers, we needed to find a general purpose way of doing this.  This is where -LdapFilter  was a life saver.  We could issue 256 calls like

-LdapFilter "(objectGUID=\00*)"
…
-LdapFilter "(objectGUID=\ff*)"

We knew that every user had to have this field and it wouldn’t matter if the user was in US, Germany or Japan.  We knew that we would capture every user.  We started with LastName -like ‘a*’, but when you’re dealing with different character sets, you can’t just run through a – z and be assured of capturing all users

Our fear is that one day, Microsoft will actually deliver on the promise that -LdapFilter will not be a part of the parameters to Get-CsOnlineUser and then we have to tell our users why they are no longer able to get users from O365 anymore

Is it possible that Microsoft would add -Skip and -Top to the Get-CsOnlineUser, so we could cycle through the entire list of users “n” at a time?

That or is there any plans on migrating or co-supporting this with GraphAPI?

Thank you

**Comment by yogkumgit** [comment]
@isrumnon

**Comment by jmgarciasanti** [comment]
@sarollins we have the same problem (100K users) and we were using same solution, a loop using "GUID like [0-9]*" filter. Since we've migrated to MicrosoftTeams module version 4.4.1, once Microsoft has stopped the service for versions <4, we're stopped at the same point.
Bad news that version 4.4.1 doesn't suppor LDAPFilter, and we neither can loop with a "UPN like [a-z]*" filter because we know we won't capture all users!

@isrumnon is there any plan to release a module supporting "Identity like...." filter?
Regards.

**Comment by isrumnon** [comment]
Hi @jmgarciasanti yes we plan on introducing in the near future (3 - 4 months from now)

**Comment by isrumnon** [comment]
Also @jmgarciasanti  on your other issues please reach out to me on isrumnon@microsoft.com 
Deleting this post due to privacy reasons

**Comment by yogkumgit** [comment]
@isrumnon Thank you very much for sharing this explanation. @sarollins Hope this comment is helpful for you. If you see a documentation update is required, please feel free to open an issue for the same. We proceed here to close it. Thanks for taking out some time to open the issue. Appreciate and encourage you to do the same in future also.
